includes/REST_API/URL_Controller.php

```diff
@@ -0,0 +1,245 @@
+ */
+
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\REST_API;
+
+use Google\Web_Stories\Story_Post_Type;
+use Google\Web_Stories\Traits\Post_Type;
+use Google\Web_Stories\Traits\Types;
+use WP_Error;
+use WP_Http;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * API endpoint to allow parsing of metadata from a URL
+ *
+ * Class URL_Controller
+ */
+class URL_Controller extends REST_Controller {
+	use Post_Type, Types;
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'url';
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @since 1.11.0
+	 *
+	 * @see register_rest_route()
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'parse_url' ],
+					'permission_callback' => [ $this, 'parse_url_permissions_check' ],
+					'args'                => [
+						'url' => [
+							'description' => __( 'The URL to process.', 'web-stories' ),
+							'required'    => true,
+							'type'        => 'string',
+							'format'      => 'uri',
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Parses a URL to return some metadata for inserting links.
+	 *
+	 * @since 1.11.0
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function parse_url( $request ) {
+		$url = untrailingslashit( $request['url'] );
+
+		if ( empty( $url ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		/**
+		 * Filters the link data TTL value.
+		 *
+		 * @since 1.11.0
+		 *
+		 * @param int $time Time to live (in seconds). Default is 1 day.
+		 * @param string $url The attempted URL.
+		 */
+		$cache_ttl = apply_filters( 'web_stories_url_data_cache_ttl', DAY_IN_SECONDS, $url );
+		$cache_key = 'web_stories_url_data_' . md5( $url );
+
+		$data = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			$response = $this->prepare_item_for_response( json_decode( $data, true ), $request );
+
+			return rest_ensure_response( $response );
+		}
+
+		$response = wp_safe_remote_get( $url );
```

Comment:
Since we're not actually using the body and are only interested in the headers, let's make sure this is a HEAD request accordingly.

```suggestion
		$response = wp_safe_remote_head( $url );
```